src/nupic/algorithms/SpatialPooler.cpp

```diff
@@ -925,11 +934,12 @@ void SpatialPooler::updateBookeepingVars_(bool learn) {
 }
 
 
-void SpatialPooler::calculateOverlap_(const UInt inputVector[],
+void SpatialPooler::calculateOverlap_(SDR &input,
                                       vector<UInt> &overlaps) const {
   overlaps.assign(numColumns_, 0);
-  connectedSynapses_.rightVecSumAtNZ(inputVector, inputVector + numInputs_,
-                                     overlaps.begin(), overlaps.end());
+  const auto &inputVector = input.getDense();
+  connectedSynapses_.rightVecSumAtNZ(inputVector.begin(), inputVector.end(),
```

Comment:
try harder to use sparse vectors only (=less conversions in SDR, shorter loops, better memory), 
see `math/SparseBinaryMatrix.hpp` there's also `*NZSparse()` variant of this method!

Reply:
I found two problems with method rightVecSumAtNZSparse:
1) class SparseBinaryMatrix does not have this method, only SparseMatrix has it.
2) The source code for rightVecSumAtNZSparse says the following:
math/SparseMatrix.hpp Line 10240:
```
    // One approach would be to walk a sorted list of ones for every row in the
    // matrix. But x is queried for every nonzero in the matrix, so in practice
    // it's much faster to create an x dense array and then perform simple
    // lookups on the dense array.
    std::fill(indb_, indb_ + nCols(), (size_type)0);
    for (InputIterator x_one = x_ones_begin; x_one != x_ones_end; ++x_one) {
      indb_[*x_one] = 1;
    }
```

Reply:
I have cloned your branch and will have a look